### PR TITLE
Made the code of a validation example more robust

### LINKED ISSUE #9979: [Validator] Encourage type safety
Example provided at https://symfony.com/doc/current/validation/custom_constraint.html#creating-the-validator-itself shows example validation as:

```php
    public function validate($value, Constraint $constraint)
    {
        if (!preg_match('/^[a-zA-Z0-9]+$/', $value, $matches)) {
            $this->context->buildViolation($constraint->message)
                ->setParameter('{{ string }}', $value)
                ->addViolation();
        }
    }
```

This is a perfectly fine example, but sets a trap creating assumption that `$value` is always a string. I think at minimum we should include:

```php
        if (!is_string($value)) {
            throw new UnexpectedTypeException($value, 'string');
        }
```

if not even something like `DateTimeValidator` has:

```php
        if (!is_scalar($value) && !(is_object($value) && method_exists($value, '__toString'))) {
            throw new UnexpectedTypeException($value, 'string');
        }

        $value = (string) $value;
```

WDYT?

### PR DESCRIPTION
**Pull request by javiereguiluz**
Fixes #9979.

### CODE REVIEW

File: validation/custom_constraint.rst [hunk 1]
```diff
@@ -59,11 +59,16 @@ The validator class is also simple, and only has one required method ``validate(
 
     use Symfony\Component\Validator\Constraint;
     use Symfony\Component\Validator\ConstraintValidator;
+    use Symfony\Component\Validator\Exception\UnexpectedTypeException;
 
     class ContainsAlphanumericValidator extends ConstraintValidator
     {
         public function validate($value, Constraint $constraint)
         {
+            if (!is_string($value) && !(is_object($value) && method_exists($value, '__toString'))) {
```
Comment (HeahDude) [inline validation/custom_constraint.rst]: Now that =I review this, I've got a comment :). If a constraint allows casting an object to null then it should do so for any scalar. What about changing `is_string` to `is_scalar`? I can see that it's done here for example https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Validator/Constraints/EmailValidator.php#L77.

Also by convention; a constraint should not test empty value, for this a `NotNull` or `NotBlank` constraint should be used. So it misses:
```php
if (null === $value || '' === $value) {
    return;
}
```
at first.

Comment (javiereguiluz) [reply validation/custom_constraint.rst]: I disagree with the proposed `is_scalar()` change. Even the error message of the validator says: `The string "{{ string }}" contains an illegal character ...`  so this must only work with strings, right?

Comment (javiereguiluz) [reply validation/custom_constraint.rst]: I've added this:

```php
if ('' === $value) {
    return;
}
```

Should it contain a `trim($value)` call too? Thanks!

Comment (HeahDude) [reply validation/custom_constraint.rst]: Hmm, casting should be ok, it's about changing the representation to make it fit, then throw an exception if it's not possible. I think we should stay consistent, if we want to keep `is_string` that strict, then let's remove the `__toString()` check.

Trimming is not as it alters the given value. Fortunately this will soon be opt-in https://github.com/symfony/symfony/pull/26484.

File: validation/custom_constraint.rst [hunk 2]
```diff
@@ -69,6 +69,10 @@ The validator class is also simple, and only has one required method ``validate(
                 throw new UnexpectedTypeException($value, 'string');
             }
 
+            if ('' === $value) {
```
Comment (HeahDude) [inline validation/custom_constraint.rst]: `null` must be check too.

Comment (javiereguiluz) [reply validation/custom_constraint.rst]: But the `null` value would never be checked. We throw an exception if `!is_string()` above.

Comment (HeahDude) [reply validation/custom_constraint.rst]: This check must be moved before indeed, and test null.

Comment (HeahDude) [reply validation/custom_constraint.rst]: The reason is that the constraint should not throw on invalid empty values, or it would superseed other constraints responsibility.

File: validation/custom_constraint.rst (hunk 1 above)
Comment (javiereguiluz) [reply validation/custom_constraint.rst]: @HeahDude I made more changes according to your comments. Thanks for the reviews and for your patience 🙏

**Review by HeahDude — APPROVED** [review]
Thanks for your work on this!

File: validation/custom_constraint.rst [hunk 3]
```diff
@@ -59,11 +59,20 @@ The validator class is also simple, and only has one required method ``validate(
...
     use Symfony\Component\Validator\Constraint;
     use Symfony\Component\Validator\ConstraintValidator;
+    use Symfony\Component\Validator\Exception\UnexpectedTypeException;
 
     class ContainsAlphanumericValidator extends ConstraintValidator
     {
         public function validate($value, Constraint $constraint)
         {
+            if (null === $value || '' === $value) {
+                return;
```
Comment (HeahDude) [inline validation/custom_constraint.rst]: What about adding a commented code line for putting the reason why we must do so?
